File: backend/quant_core/cashflow.py
import calendar
from datetime import date
from typing import List, Dict, Any

def subtract_months(dt: date, months: int) -> date:
    """
    Subtracts a number of months from a date, preserving the day of the month 
    (or capping it at the maximum day of the target month).
    """
    month = dt.month - 1 - months
    year = dt.year + month // 12
    month = month % 12 + 1
    day = min(dt.day, calendar.monthrange(year, month)[1])
    return date(year, month, day)

def generate_coupon_dates(issue_date: date, maturity_date: date, coupon_frequency: int = 2) -> List[date]:
    """
    Generates theoretical coupon dates by counting backwards from maturity_date.
    Returns [t_0, t_1, ..., t_N] where:
      - t_0 is the preceding theoretical coupon date (on or before issue_date)
      - t_1 is the first coupon payment date
      - t_N is the maturity_date
    """
    dates = []
    months_per_period = 12 // coupon_frequency
    k = 0
    while True:
        # Subtracting from the original maturity_date to avoid cumulative day shifting
        dt = subtract_months(maturity_date, k * months_per_period)
        dates.append(dt)
        if dt <= issue_date:
            break
        k += 1
        
    dates.reverse()
    return dates

def generate_cashflows(
    issue_date: date,
    maturity_date: date,
    coupon_rate: float,
    coupon_frequency: int = 2,
    face_value: float = 100.0
) -> List[Dict[str, Any]]:
    """
    Generates the explicit cashflow schedule for a G-Sec.
    Correctly accounts for odd first coupon periods by prorating the coupon amount.
    """
    dates = generate_coupon_dates(issue_date, maturity_date, coupon_frequency)
    if len(dates) < 2:
        # Fallback for immediate/invalid schedules
        return [{"date": maturity_date, "amount": face_value, "type": "redemption"}]
        
    cashflows = []
    regular_coupon = face_value * (coupon_rate / 100.0) / coupon_frequency
    
    for i in range(1, len(dates)):
        pay_date = dates[i]
        
        if i == 1:
            t_0 = dates[0]
            t_1 = dates[1]
            # Prorate the first coupon if issue_date falls after the preceding theoretical coupon date t_0
            if issue_date > t_0:
                period_days = (t_1 - t_0).days
                accrued_days = (t_1 - issue_date).days
                amount = regular_coupon * (accrued_days / period_days) if period_days > 0 else 0.0
            else:
                amount = regular_coupon
        else:
            amount = regular_coupon
            
        is_last = (i == len(dates) - 1)
        
        if is_last:
            cashflows.append({
                "date": pay_date,
                "amount": amount + face_value,
                "type": "both"
            })
        else:
            cashflows.append({
                "date": pay_date,
                "amount": amount,
                "type": "coupon"
            })
            
    return cashflows

File: lib/pricing-engine/cashflow.ts
import { Cashflow } from "./types";

export function subtractMonths(dt: Date, months: number): Date {
  const year = dt.getFullYear();
  const month = dt.getMonth(); // 0-indexed
  const day = dt.getDate();
  
  const targetMonthTotal = year * 12 + month - months;
  const targetYear = Math.floor(targetMonthTotal / 12);
  const targetMonth = targetMonthTotal % 12;
  
  // Get last day of target year and month
  const maxDays = new Date(targetYear, targetMonth + 1, 0).getDate();
  const targetDay = Math.min(day, maxDays);
  
  return new Date(targetYear, targetMonth, targetDay);
}

export function generateCouponDates(
  issueDate: Date, 
  maturityDate: Date, 
  couponFrequency: number = 2
): Date[] {
  const dates: Date[] = [];
  const monthsPerPeriod = 12 / couponFrequency;
  let k = 0;
  
  while (true) {
    const dt = subtractMonths(maturityDate, k * monthsPerPeriod);
    dates.push(dt);
    if (dt.getTime() <= issueDate.getTime()) {
      break;
    }
    k++;
  }
  
  dates.reverse();
  return dates;
}

export function generateCashflows(
  issueDate: Date,
  maturityDate: Date,
  couponRate: number,
  couponFrequency: number = 2,
  faceValue: number = 100.0
): Cashflow[] {
  const dates = generateCouponDates(issueDate, maturityDate, couponFrequency);
  if (dates.length < 2) {
    return [{ date: maturityDate, amount: faceValue, type: "redemption" }];
  }
  
  const cashflows: Cashflow[] = [];
  const regularCoupon = faceValue * (couponRate / 100.0) / couponFrequency;
  
  for (let i = 1; i < dates.length; i++) {
    const payDate = dates[i];
    let amount = regularCoupon;
    
    if (i === 1) {
      const t0 = dates[0];
      const t1 = dates[1];
      if (issueDate.getTime() > t0.getTime()) {
        const periodDays = Math.round((t1.getTime() - t0.getTime()) / (1000 * 60 * 60 * 24));
        const accruedDays = Math.round((t1.getTime() - issueDate.getTime()) / (1000 * 60 * 60 * 24));
        amount = periodDays > 0 ? regularCoupon * (accruedDays / periodDays) : 0.0;
      }
    }
    
    const isLast = (i === dates.length - 1);
    if (isLast) {
      cashflows.push({
        date: payDate,
        amount: amount + faceValue,
        type: "both"
      });
    } else {
      cashflows.push({
        date: payDate,
        amount: amount,
        type: "coupon"
      });
    }
  }
  
  return cashflows;
}
